integration/format/format.go

```diff
@@ -0,0 +1,175 @@
+
+	ts := memory.NewTreeStorage()
+	ls := memory.NewLogStorage(ts, monitoring.InertMetricFactory{})
+	as := memory.NewAdminStorage(ts)
+	tree, err := createTree(ctx, as, ls)
+	if err != nil {
+		return "", err
+	}
+	log.InitMetrics(nil)
+
+	leaves := generateLeaves(treeSize, leafFormat)
+	if err := sequenceLeaves(ctx, ls, tree, leaves, batchSize); err != nil {
+		return "", err
+	}
+
+	// Read the latest LogRoot back.
+	var root types.LogRootV1
+	if err := ls.ReadWriteTransaction(ctx, tree, func(ctx context.Context, tx storage.LogTreeTX) error {
+		latest, err := tx.LatestSignedLogRoot(ctx)
+		if err != nil {
+			return err
+		}
+		return root.UnmarshalBinary(latest.LogRoot)
+	}); err != nil {
+		return "", fmt.Errorf("ReadWriteTransaction: %v", err)
+	}
+
+	return latestRevisions(ls, tree.TreeId, rfc6962.DefaultHasher)
+}
+
+func createTree(ctx context.Context, as storage.AdminStorage, ls storage.LogStorage) (*trillian.Tree, error) {
+	tree, err := storage.CreateTree(ctx, as, &trillian.Tree{
+		TreeType:        trillian.TreeType_LOG,
+		TreeState:       trillian.TreeState_ACTIVE,
+		MaxRootDuration: durationpb.New(0 * time.Millisecond),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("CreateTree: %v", err)
+	}
+
+	logRoot, err := (&types.LogRootV1{RootHash: rfc6962.DefaultHasher.EmptyRoot()}).MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("MarshalBinary: %v", err)
+	}
+
+	if err = ls.ReadWriteTransaction(ctx, tree, func(ctx context.Context, tx storage.LogTreeTX) error {
+		return tx.StoreSignedLogRoot(ctx, &trillian.SignedLogRoot{LogRoot: logRoot})
+	}); err != nil {
+		return nil, fmt.Errorf("ReadWriteTransaction: %v", err)
+	}
+
+	return tree, nil
+}
+
+func generateLeaves(count int, format string) []*trillian.LogLeaf {
+	leaves := make([]*trillian.LogLeaf, 0, count)
+	for i := 0; i < count; i++ {
+		data := []byte(fmt.Sprintf(format, i))
+		hash := sha256.Sum256(data)
+		leaves = append(leaves, &trillian.LogLeaf{
+			LeafValue: data, LeafIdentityHash: hash[:], MerkleLeafHash: hash[:],
+		})
+	}
+	return leaves
+}
+
+func sequenceLeaves(ctx context.Context, ls storage.LogStorage, tree *trillian.Tree, leaves []*trillian.LogLeaf, batchSize int) error {
+	for i, size := 0, len(leaves); i < size; i += batchSize {
+		if left := size - i; left < batchSize {
+			batchSize = left
+		}
+		if _, err := ls.QueueLeaves(ctx, tree, leaves[i:i+batchSize], time.Now()); err != nil {
+			return fmt.Errorf("QueueLeaves: %v", err)
+		}
+
+		sequenced, err := log.IntegrateBatch(ctx, tree, batchSize, 0, 24*time.Hour, clock.System, ls, quota.Noop())
+		if err != nil {
+			return fmt.Errorf("IntegrateBatch: %v", err)
+		}
+		if got, want := sequenced, batchSize; got != want {
+			return fmt.Errorf("IntegrateBatch: got %d, want %d", got, want)
+		}
+	}
+	return nil
+}
+
+type treeAndRev struct {
+	subtree  *storagepb.SubtreeProto
+	revision int
+}
+
+func latestRevisions(ls storage.LogStorage, treeID int64, hasher merkle.LogHasher) (string, error) {
```

Comment:
What do the revisions here refer to?

Reply:
In the original Trillian storage design the Merkle tree is split into tiles (which are called `SubtreeProto` there), and each tile will have a new revision every time it's modified (when new entries are appended to the log, and some nodes of a tile get new hashes, or modified ephemeral hashes).

Trillian also had maps, which were sparse Merkle trees. Maps had different API: it was possible to modify an arbitrary location in the Merkle tree (as opposed to logs, where only appends to the end are allowed). Trillian tried to have a unified Merkle storage (the `SubtreeProto`s with revisions, etc). Since then maps were deleted and the design of logs converges to simpler version which does not store ephemeral nodes, and does not need tile revisions. But for backwards compatibility the revisions still exist.

Reply:
A revision is a "snapshot" when a new tree head is created. For logs, a new revision is when new batch of entries are appended and committed to storage.

Reply:
So, the way to see the state of Merkle tree at revision `x`, one should scan all tiles/`SubtreeProto`s, and take the latest revision <= `x` for each.

Reply:
Thank you, that elucidated my understanding a lot!